### ISSUE TITLE
There are still data races in SDA core

### ISSUE THREAD

**Issue by Liamsi** [dedis/cothority#240]
Please run your simulations with the `-race` detection on *and* write more tests. 
```
WARNING: DATA RACE
Read by goroutine 13:
  github.com/dedis/cothority/lib/network.(*SecureTcpConn).String()
      <autogenerated>:26 +0x81
  fmt.(*pp).handleMethods()
      /usr/local/Cellar/go/1.6/libexec/src/fmt/print.go:730 +0x751
  fmt.(*pp).printArg()
      /usr/local/Cellar/go/1.6/libexec/src/fmt/print.go:806 +0x5ba
  fmt.(*pp).doPrint()
      /usr/local/Cellar/go/1.6/libexec/src/fmt/print.go:1273 +0x3fc
  fmt.Sprintln()
      /usr/local/Cellar/go/1.6/libexec/src/fmt/print.go:271 +0x6f
  github.com/dedis/cothority/lib/dbg.Lvl()
     
dedis/cothority/lib/dbg/debug_lvl.go:84 +0x7f3
  github.com/dedis/cothority/lib/dbg.Lvld()
     dedis/cothority/lib/dbg/debug_lvl.go:53 +0x4c
  github.com/dedis/cothority/lib/dbg.Lvl3()
     dedis/cothority/lib/dbg/debug_lvl.go:151 +0x4b
  github.com/dedis/cothority/lib/sda.(*Host).Close()
     dedis/cothority/lib/sda/host.go:240 +0x99f
  github.com/dedis/cothority/lib/sda.(*Node).CloseHost()
     dedis/cothority/lib/sda/node.go:441 +0x64
  github.com/dedis/cothority/protocols/manage.(*ProtocolCloseAll).FuncClose()
     dedis/cothority/protocols/manage/close_all.go:78 +0x798
  github.com/dedis/cothority/protocols/manage.(*ProtocolCloseAll).FuncClose-fm()
     dedis/cothority/protocols/manage/close_all.go:47 +0x4b
  runtime.call32()
      /usr/local/Cellar/go/1.6/libexec/src/runtime/asm_amd64.s:472 +0x3d
  reflect.Value.Call()
      /usr/local/Cellar/go/1.6/libexec/src/reflect/value.go:303 +0xcd
  github.com/dedis/cothority/lib/sda.(*Node).DispatchHandler()
     dedis/cothority/lib/sda/node.go:268 +0x91f
  github.com/dedis/cothority/lib/sda.(*Node).DispatchMsg()
     dedis/cothority/lib/sda/node.go:348 +0x12d2
  github.com/dedis/cothority/lib/sda.(*Overlay).TransmitMsg()
     dedis/cothority/lib/sda/overlay.go:88 +0x1096
  github.com/dedis/cothority/lib/sda.(*Host).processMessages()
     dedis/cothority/lib/sda/host.go:318 +0x490

Previous write by goroutine 12:
  github.com/dedis/cothority/lib/network.(*SecureTcpHost).Listen()
     dedis/cothority/lib/network/net.go:340 +0x54c
  github.com/dedis/cothority/lib/sda.(*Host).listen.func2()
     dedis/cothority/lib/sda/host.go:175 +0x221

Goroutine 13 (running) created at:
  github.com/dedis/cothority/lib/sda.(*Host).StartProcessMessages()
     dedis/cothority/lib/sda/host.go:292 +0x87
  main.main()
     dedis/cothority/simul/cothority/cothority.go:65 +0x8c5

Goroutine 12 (running) created at:
  github.com/dedis/cothority/lib/sda.(*Host).listen()
     dedis/cothority/lib/sda/host.go:179 +0x381
```

**Comment by ineiti** [comment]
what branch? what test?

**Comment by ineiti** [comment]
That should be a BUG then, no?

**Comment by Liamsi** [comment]
> what branch? what test?

Short version: on sda_1 branch while running any handler related simulation (no test). Long version: On my WIP branch for naive/ntree (#200 #201). But as you can see from the output the race is not related to that code. But rather to the current sda code using handlers. You get even the same errors when running the handlers example simulation (on the sda_1 branch). 
Could be a simple debug output related data race.

**Comment by Liamsi** [comment]
> That should be a BUG then, no?

Of course.

**Comment by Liamsi** [comment]
Merged with #262
